### PR TITLE
feat: Add CRUD apis for role, permission, user

### PR DESCRIPTION
**Pull request by mayurnewase**
### Description

Add permission, role, user CRUD apis.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Is CRUD MVC related.
- [x] Is Auth, RBAC security related.
- [ ] Changes the security db schema.
- [ ] Introduces new feature
- [ ] Removes existing feature

### CODE REVIEW

File: flask_appbuilder/tests/test_api.py [hunk 1]
```diff
@@ -3100,3 +3100,350 @@ def test_before_request(self):
...
+        self.app.config.from_object("flask_appbuilder.tests.config_api")
+        self.app.config["ENABLE_USER_CRUD_API"] = True
+        self.db = SQLA(self.app)
+        self.appbuilder = AppBuilder(self.app, self.db.session)
+
+        # TODO: this heinous hack is to avoid using stale db session leaking from
+        # RolePermissionAPITestCase
+        # don't know why all baseviews in Appbuilder are attached to stale session,
+        # causing error when adding a new user which reads roles from this session and
+        # datamodel uses stale session to add it.
```
Comment (mayurnewase) [inline flask_appbuilder/tests/test_api.py:3121]: Don't really know why this happens.
If I run this test before `RolePermissionAPITestCase` it works fine, but if I run `RolePermissionAPITestCase` first then `datamodel` objects in the new `modelRestApi` classes already have session attached, so this line https://github.com/dpgaspar/Flask-AppBuilder/blob/8a18281002d55a50416cf98dad8d75c4c218dbb8/flask_appbuilder/base.py#L343 skips assigning current session to it.

File: flask_appbuilder/security/sqla/apis/user/user.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+class UserApi(ModelRestApi):
+    resource_name = "user"
+    class_permission_name = "User"
+    datamodel = SQLAInterface(User)
+    allow_browser_login = True
+
+    list_columns = [
+        "id",
+        "roles",
```
Comment (dpgaspar) [inline flask_appbuilder/security/sqla/apis/user/user.py:26]: try: `roles.id`, `roles.name`

File: flask_appbuilder/security/sqla/apis/user/user.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    ]
+    edit_columns = add_columns
+    search_columns = list_columns
+
+    add_model_schema = UserPostSchema()
+    edit_model_schema = UserPutSchema()
+
+    def pre_update(self, item):
+        item.changed_on = datetime.now()
+        item.changed_by_fk = g.user.id
```
Comment (dpgaspar) [inline flask_appbuilder/security/sqla/apis/user/user.py:60]: isn't this getting handled by the Audit mixin?

File: flask_appbuilder/security/sqla/apis/user/user.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+              $ref: '#/components/responses/401'
+            404:
+              $ref: '#/components/responses/404'
+            422:
+              $ref: '#/components/responses/422'
+            500:
+              $ref: '#/components/responses/500'
+        """
+        try:
+            item = self.add_model_schema.load(request.json)
```
Comment (dpgaspar) [inline flask_appbuilder/security/sqla/apis/user/user.py:101]: check https://flask-appbuilder.readthedocs.io/en/latest/rest_api.html#validation-and-custom-validation. Try BaseModelSchema instead of marshmallow schemas, it will give you the SQLAlchemy model for free.

File: flask_appbuilder/security/sqla/apis/user/user.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            for key, value in item.items():
+                if key != "roles":
+                    setattr(model, key, value)
+                else:
+                    for role_id in item[key]:
+                        role = (
+                            current_app.appbuilder.get_session.query(Role)
+                            .filter(Role.id == role_id)
+                            .first()
+                        )
```
Comment (dpgaspar) [inline flask_appbuilder/security/sqla/apis/user/user.py:113]: if role is None?

File: flask_appbuilder/tests/test_api.py [hunk 6]
```diff
@@ -3100,3 +3100,350 @@ def test_before_request(self):
...
+            "roles": [1],
+            "username": "fasab_usear_api_test",
+        }
+        rv = self.auth_client_post(client, token, uri, create_user_payload)
+        add_user_response = json.loads(rv.data)
+        self.assertEqual(rv.status_code, 201)
+        assert "id" and "result" in add_user_response
+        self.assertEqual(create_user_payload, add_user_response["result"])
+
+        uri = f"api/v1/user/{add_user_response['id']}"
```
Comment (dpgaspar) [inline flask_appbuilder/tests/test_api.py:3161]: let's test the crud on separate tests. One to get list, one got get item, one for non existent get item, one for create user, a couple for create user edge cases (no roles, invalid password etc), one for update and update edge cases/error conditions

File: flask_appbuilder/tests/test_api.py [hunk 7]
```diff
@@ -3100,3 +3100,350 @@ def test_before_request(self):
...
+
+        uri = "api/v1/permission/"
+        create_permission_payload = {"name": "super duper fab permission"}
+        rv = self.auth_client_post(client, token, uri, create_permission_payload)
+        add_permission_response = json.loads(rv.data)
+        self.assertEqual(rv.status_code, 201)
+        assert "id" and "result" in add_permission_response
+        self.assertEqual(create_permission_payload, add_permission_response["result"])
+
+        uri = f"api/v1/permission/{add_permission_response['id']}"
```
Comment (dpgaspar) [inline flask_appbuilder/tests/test_api.py:3216]: let's split the tests, and add more for error conditions and edge cases

File: flask_appbuilder/tests/test_api.py [hunk 8]
```diff
@@ -3100,3 +3100,350 @@ def test_before_request(self):
         uri = "api/v1/model1beforerequest/update"
```
Comment (dpgaspar) [inline flask_appbuilder/tests/test_api.py:3100]: I think it makes to create a new test file for this: `tests/test_security_api.py` for example

**Review by dpgaspar — COMMENTED** [review]
Looking very good

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 4 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/user.py:101]: but that won't work as we need to populate roles right?

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 3 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/user.py:60]: the default `User` model doesn't use Auditmixin and have `changed_on` column defined without `onupdate` property.
so had to add this.

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 5 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/user.py:113]: actually there is a validator in schema, to check length at least 1 and required true so I assumed it won't be null or empty list for POST /user.

File: flask_appbuilder/tests/test_api.py (hunk 8 above)
Comment (mayurnewase) [reply flask_appbuilder/tests/test_api.py]: done.

File: flask_appbuilder/tests/test_api.py (hunk 6 above)
Comment (mayurnewase) [reply flask_appbuilder/tests/test_api.py]: done.

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 5 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/user.py]: handled and added test.

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 4 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/user.py]: but we need to populate roles list first before assiging to model right?

File: flask_appbuilder/security/sqla/apis/permission_view_menu/api.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+from flask_appbuilder.security.sqla.models import PermissionView
+
+
+class PermissionViewMenuApi(ModelRestApi):
+    resource_name = "permissionviewmenu"
+    class_permission_name = "PermissionViewMenu"
+    datamodel = SQLAInterface(PermissionView)
+    allow_browser_login = True
+
+    list_columns = ["id", "permission_id", "view_menu_id"]
```
[the remark this answers is not in the document]
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/permission_view_menu/api.py]: done.

File: flask_appbuilder/security/sqla/apis/user/schema.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+from marshmallow import fields, Schema
+from marshmallow.validate import Length
+
+from .validator import PasswordComplexityValidator
+
+
+class UserPostSchema(Schema):
+    active = fields.Boolean(default=False)
```
[the remark this answers is not in the document]
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/schema.py]: added description.
default was set according to UI, which has active field unchecked by default, but I updated this to active by default

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 2 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/user.py]: done.

File: flask_appbuilder/tests/test_api.py (hunk 7 above)
Comment (mayurnewase) [reply flask_appbuilder/tests/test_api.py]: done.

**Comment by mayurnewase** [comment]
my sql test is bahaving wierd.
its straight forward read api, and it't not able to find the permission which security manager added.

File: flask_appbuilder/security/sqla/apis/user/api.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        "active",
+        "email",
+        "last_login",
+        "login_count",
+        "fail_login_count",
+        "created_on",
+        "changed_on",
+        "created_by.id",
+        "changed_by.id",
+        "created.id",
```
Comment (dpgaspar) [inline flask_appbuilder/security/sqla/apis/user/api.py:40]: consider removing the `created` and `changed` backref's for performance reasons. Not relevant info IMO

**Review by dpgaspar — COMMENTED** [review]
Looking good, gave it another quick review and tried things out. ping me if you have any blockers

File: flask_appbuilder/security/sqla/apis/user/api.py (hunk 11 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/api.py:40]: I thought by default all info possible and let user specify only they columns they need.

File: flask_appbuilder/security/sqla/apis/permission_view_menu/api.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+from flask_appbuilder import ModelRestApi
+from flask_appbuilder.models.sqla.interface import SQLAInterface
+from flask_appbuilder.security.sqla.models import PermissionView
+
+
+class PermissionViewMenuApi(ModelRestApi):
+    resource_name = "permissionviewmenu"
```
[the remark this answers is not in the document]
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/permission_view_menu/api.py:7]: i think able to manage permission to views mapping is useful.

File: flask_appbuilder/security/sqla/apis/role/api.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+from flask_appbuilder.security.sqla.models import Role
+
+
+class RoleApi(ModelRestApi):
+    resource_name = "role"
+    class_permission_name = "Role"
+    datamodel = SQLAInterface(Role)
+    allow_browser_login = True
+
+    list_columns = ["id", "permissions", "name"]
```
[the remark this answers is not in the document]
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/role/api.py:12]: makes sense, we can give more info in this api like permission name and view_menu name.

File: flask_appbuilder/security/sqla/apis/role/api.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+    show_columns = list_columns
+    add_columns = ["name", "permissions"]
+    edit_columns = ["name", "permissions"]
```
[the remark this answers is not in the document]
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/role/api.py:15]: how about we just keep 1 api POST role/{pk}/permission to set permissions on role without PUT api?

File: flask_appbuilder/security/sqla/manager.py [hunk 15]
```diff
@@ -84,6 +92,13 @@ def get_session(self):
         return self.appbuilder.get_session
 
     def register_views(self):
+        if self.appbuilder.app.config.get("ENABLE_USER_CRUD_API", False):
```
Comment (mayurnewase) [inline flask_appbuilder/security/sqla/manager.py:95]: now as we have renamed these apis as security apis, I wonder if we should just reuse existing `FAB_ADD_SECURITY_VIEWS` for this too.

File: flask_appbuilder/security/sqla/apis/user/user.py (hunk 3 above)
Comment (dpgaspar) [reply flask_appbuilder/security/sqla/apis/user/user.py]: no, your right

File: flask_appbuilder/security/sqla/apis/user/api.py (hunk 11 above)
Comment (dpgaspar) [reply flask_appbuilder/security/sqla/apis/user/api.py]: I think it's dangerous, on dbs with lot's of users

File: flask_appbuilder/tests/test_security_api.py [hunk 16]
```diff
@@ -0,0 +1,1097 @@
...
+        token = self.login(client, USERNAME_ADMIN, PASSWORD_ADMIN)
+
+        num = 3
+        role_name = f"test_edit_role_api_{num}"
+        role_2_name = f"test_edit_role_api_{num+1}"
+        permission_1_name = f"test_edit_role_permission_{num}"
+        permission_2_name = f"test_edit_role_permission_{num+1}"
+        view_menu_name = f"test_edit_role_view_menu_{num}"
+
+        # permission_1_view_menu = self.appbuilder.sm.add_permission_view_menu(
```
Comment (dpgaspar) [inline flask_appbuilder/tests/test_security_api.py:723]: delete?

**Review by dpgaspar — COMMENTED** [review]
Looking great! left a couple of comments, thank you for your patience ;)

File: flask_appbuilder/security/sqla/apis/user/api.py (hunk 11 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/apis/user/api.py]: agree, removed already.

File: flask_appbuilder/security/sqla/manager.py (hunk 15 above)
Comment (mayurnewase) [reply flask_appbuilder/security/sqla/manager.py:95]: agree, and it also makes it consistent with FAB feature flags.

File: flask_appbuilder/security/sqla/apis/user/api.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+                            role.user_id = model.id
+                            role.role_id = role_id
+                            roles.append(role)
+
+            if "roles" in item.keys():
+                model.roles = roles
+
+            self.pre_add(model)
+            self.datamodel.add(model, raise_exception=True)
+            # return self.response(201, id=model.id)
```
Comment (dpgaspar) [inline flask_appbuilder/security/sqla/apis/user/api.py:129]: delete?

**Review by dpgaspar — APPROVED** [review]
